driver/src/orderbook/shadow_orderbook.rs

```diff
@@ -0,0 +1,295 @@
+                    shadow: shadow_balance,
+                })
+            }
+        }
+
+        changes
+    }
+}
+
+/// Represents a change in order data between a primary and shadow orderbook.
+#[derive(Debug, PartialEq)]
+enum OrderChange {
+    /// An order was added, i.e. it exists in the primary but not the shadow
+    /// orderbook.
+    Added(Order),
+    /// An order was removed, i.e. it is missing from the primary orderbook but
+    /// exists in shadow orderbook.
+    Removed(Order),
+    /// An order was modified, i.e. it exists in both orderbooks but with
+    /// different values.
+    Modified {
+        user: Address,
+        id: u16,
+        primary: OrderValues,
+        shadow: OrderValues,
+    },
+}
+
+impl OrderChange {
+    /// Compares a primary and shadow orderbook order vector and return the
+    /// detected changes.
+    fn compare_orders(primary: &[Order], shadow: &[Order]) -> Vec<OrderChange> {
+        let mut shadow_orders = shadow
+            .iter()
+            .map(|order| ((order.account_id, order.id), order))
+            .collect::<HashMap<_, _>>();
+
+        let mut changes = Vec::new();
+        for primary_order in primary {
+            let (user, id) = (primary_order.account_id, primary_order.id);
+            let change = match shadow_orders.remove(&(user, id)) {
+                None => OrderChange::Added(primary_order.clone()),
+                Some(shadow_order) if primary_order != shadow_order => OrderChange::Modified {
+                    user,
+                    id,
+                    primary: OrderValues::from(primary_order),
+                    shadow: OrderValues::from(shadow_order),
+                },
+                _ => continue,
+            };
+            changes.push(change);
+        }
+
+        for (_, remaining_shadow_order) in shadow_orders {
+            changes.push(OrderChange::Removed(remaining_shadow_order.clone()))
+        }
+
+        changes
+    }
+
+    /// Retrieves the user and order ID for the order change.
+    #[cfg(test)]
+    fn user_and_id(&self) -> (Address, u16) {
+        match self {
+            OrderChange::Added(order) | OrderChange::Removed(order) => (order.account_id, order.id),
+            OrderChange::Modified { user, id, .. } => (*user, *id),
+        }
+    }
+}
+
+/// Values that can possibly differ between orders.
+#[derive(Debug, PartialEq)]
+struct OrderValues {
+    buy_token: u16,
+    sell_token: u16,
+    buy_amount: u128,
+    sell_amount: u128,
+}
+
+impl From<&'_ Order> for OrderValues {
+    fn from(order: &Order) -> Self {
+        OrderValues {
+            buy_token: order.buy_token,
+            sell_token: order.sell_token,
+            buy_amount: order.buy_amount,
+            sell_amount: order.sell_amount,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn account_state_diff() {
+        let addr = |i: u8| Address::repeat_byte(i);
```

Comment:
my first accurately labeled UBERNIT: 

Any particular reason the bytes are chosen to be repeated? Is this the simplest method offered by Address to construct something quick? I just wonder because I guess this makes `addr(1) = 0x11111.11111`  rather than `0x0000...1` so if someone is expected to get the "1-Address", they would be confused to discover that it is not `addr(1)`

Reply:
It would make it `0x010101...01`. I had used `Address::from_low_u64_le(number)` but when seeing the `assert_eq` failure messages, it was very hard to keep track of the addresses since they were mostly 0's and with just 1 important number at the very end. It was especially hard when the 0's were split with a newline because of terminal width.

Maybe I can rename the closure to something more descriptive? But in general I do prefer using `repeat_byte` here as it makes the assertion failures easier to read.

Reply:
Nah, this seems like pretty good rational to me!